### PR TITLE
Add GDALRelationship class for describing a relationship between two tables, implement for some drivers

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR adds a new GDALRelationship class for describing the relationship between two tables.

This class is designed to be somewhat of a superset of the relationship properties utilised by formats which are likely to gain support for the relationship API in future, specifically Geopackage (via the related tables extension), FileGeodatabase and the various database drivers (postgres/sql server/oracle/...).

This PR also implements relationship discovery in the GPKG, openfilegdb, filegdb and pgeo drivers

### CODE REVIEW

**Comment by nyalldawson** [comment]
@rouault this is ready for review/merge

File: ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp [hunk 1]
```diff
@@ -1873,6 +1879,82 @@ void GDALGeoPackageDataset::FixupWrongMedataReferenceColumnNameUpdate()
...
+    m_osMapRelationships.clear();
+
+    auto oResultTable = SQLQuery(hDB, "SELECT base_table_name, base_primary_column, "
+                                 "related_table_name, related_primary_column, relation_name, "
+                                 "mapping_table_name FROM gpkgext_relations" );
+    if ( oResultTable && oResultTable->RowCount() > 0 )
+    {
+        for(int i=0; i<oResultTable->RowCount();i++)
+        {
+            const char* pszBaseTableName = oResultTable->GetValue(0, i);
```
Comment (rouault) [inline ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp:1898]: for robustness, you should assume that the return value of all those GetValue() might be nullptr (corrupted DB, etc.), and thus test for it before dereferencing/passing it to other functions

File: ogr/ogrsf_frmts/openfilegdb/ogropenfilegdbdatasource.cpp [hunk 2]
```diff
@@ -526,11 +527,12 @@ int OGROpenFileGDBDataSource::OpenFileGDBv10(int iGDBItems,
     if( !oTable.Open(osFilename, false) )
         return FALSE;
 
-    int iUUID = oTable.GetFieldIdx("UUID");
-    int iName = oTable.GetFieldIdx("Name");
-    int iPath = oTable.GetFieldIdx("Path");
-    int iDefinition = oTable.GetFieldIdx("Definition");
-    int iDocumentation = oTable.GetFieldIdx("Documentation");
+    const int iUUID = oTable.GetFieldIdx("UUID");
+    const int iType = oTable.GetFieldIdx("Type");
```
Comment (rouault) [inline ogr/ogrsf_frmts/openfilegdb/ogropenfilegdbdatasource.cpp:531]: iType should be tested like other fields at line 536

File: gcore/gdal.h [hunk 3]
```diff
@@ -283,6 +283,11 @@ typedef void *GDALRasterAttributeTableH;
 /** Opaque type used for the C bindings of the C++ GDALAsyncReader class */
 typedef void *GDALAsyncReaderH;
 
+/** Opaque type used for the C bindings of the C++ GDALRelationship class
+ *  @since QGIS 3.6
```
Comment (rouault) [inline gcore/gdal.h:287]:
```suggestion
 *  @since GDAL 3.6
```

File: swig/include/Relationship.i [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  %apply (char **options) {char **};
+  char **GetLeftTableFields() {
+      return GDALRelationshipGetLeftTableFields(self);
+  }
+  %clear (char **);
+#else
+  void GetLeftTableFields(char ***pList) {
+      *pList = GDALRelationshipGetLeftTableFields(self);
+  }
+#endif
```
Comment (rouault) [inline swig/include/Relationship.i:123]:
```suggestion
%apply (char **CSL) {char **};
  char **GetLeftTableFields() {
     return GDALRelationshipGetLeftTableFields(self);
  }
%clear char **;
```
similarly for other instances that return a char** that must be CSLDestroy()'ed

File: gcore/gdal.h [hunk 5]
```diff
@@ -589,6 +594,12 @@ typedef struct GDALDimensionHS* GDALDimensionH;
  */
 #define GDAL_DCAP_FIELD_DOMAINS "DCAP_FIELD_DOMAINS"
 
+/** Capability set by drivers for formats which support reading table relationships.
+ *
+ * @since GDAL 3.6
+ */
+#define GDAL_DCAP_RELATIONSHIPS "DCAP_RELATIONSHIPS"
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply gcore/gdal.h:601]: I was planning to add a bunch of DMD strings which break down what create relationship options are supported. Eg one for supported cardinality types, one for general supported properties (the labels, related table type,..) etc.

File: gcore/gdal.h (hunk 3 above)
Comment (nyalldawson) [reply gcore/gdal.h:287]: :relaxed: oops, muscle memory!

**Comment by nyalldawson** [comment]
Thanks for the review @rouault  -- it's all addressed now

File: gcore/gdalrelationship.cpp [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+
+#include <algorithm>
+#include <cstddef>
+#include <new>
+
+#include "cpl_atomic_ops.h"
+#include "cpl_error.h"
+#include "cpl_multiproc.h"
+
+CPL_CVSID("$Id$")
```
Comment (rouault) [inline gcore/gdalrelationship.cpp:41]: All that can likely be removed:
```suggestion
```

**Comment by rouault** [comment]
do you think you could rework the commit history a bit (I kown I'm somewhat maniac about that), or maybe just squashing everything together ?

File: gcore/gdalrelationship.cpp (hunk 6 above)
Comment (nyalldawson) [reply gcore/gdalrelationship.cpp]: Done

**Comment by nyalldawson** [comment]
@rouault 

> do you think you could rework the commit history a bit (I kown I'm somewhat maniac about that), or maybe just squashing everything together ?

No problem! I've done that now

**Comment by nyalldawson** [comment]
Failing mac test is unrelated

**Pull request merged**